File: hp_final_design/firmware/fft.py
import cmath
import math
from typing import List


SCALE = 12
MAX_N_PER_FFT = 32
TWIDDLES = []


for m in range(1, round(math.log2(MAX_N_PER_FFT)) + 1):
    stage = 1 << m  # m = 2, 4, 8, ..., n
    W = cmath.exp(-2j * cmath.pi / stage)
    W_scaled = complex(round(W.real * (1<<SCALE)), round(W.imag * (1<<SCALE)))
    TWIDDLES.append(W_scaled)


def complex_mult(in1: complex, in2: complex) -> complex:
    a = int(in1.real)
    b = int(in1.imag)
    c = int(in2.real)
    d = int(in2.imag)

    return complex((a * c - b * d) >> SCALE, (a * d + b * c) >> SCALE)


def flog2(x: int) -> int:
    r = 0

    while x > 1:
        r += 1
        x = x >> 1

    return r


def bit_reverse(i: int, bits: int) -> int:
    r: int = 0
    for _ in range(bits):
        r = (r << 1) | (i & 1)
        i >>= 1
    return r


def fft(x: List[complex]) -> List[complex]:
    n = len(x)
    bits = flog2(n)

    # Step 1: bit-reversal permutation
    X = [0j] * n

    for i in range(n):
        X[bit_reverse(i, bits)] = x[i]

    # Step 2: iterative FFT stages
    for stage in range(1, bits + 1):
        m = 1 << stage  # m = 2, 4, 8, ..., n

        half = m // 2
        w_m = TWIDDLES[stage-1]

        for base in range(0, n, m):
            w = complex(1 << SCALE, 0)

            for k in range(half):
                t = complex_mult(w, X[base + k + half])

                u = X[base + k]
                X[base + k] = u + t
                X[base + k + half] = u - t

                w = complex_mult(w, w_m)

    return X


def inv_dft(x: List[complex]) -> List[complex]:
    n = len(x)
    X = [0j] * n

    for k in range(n):
        sum_val = 0j
        for t in range(n):
            angle = 2j * cmath.pi * t * k / n
            sum_val += x[t] * cmath.exp(angle)
        X[k] = sum_val / n

    return X

File: hp_final_design/firmware/fft.c
#include "fft.h"
#include "uart.h"

static Complex complex_mult(Complex a, Complex b)
{
    Complex result;

    result.real = ( (a.real * b.real) - (a.imag * b.imag) ) >> SCALE; 
    result.imag = ( (a.real * b.imag) + (a.imag * b.real) ) >> SCALE;

    return result;
}

static Complex complex_add(Complex a, Complex b)
{
    Complex result;

    result.real = a.real + b.real;
    result.imag = a.imag + b.imag;

    return result;
}

static Complex complex_sub(Complex a, Complex b)
{
    return complex_add(a, (Complex){.real = -b.real, .imag = -b.imag});
}

int flog2(int x)
{
    int r = 0;

    while (x > 1) {
        x = x >> 1;
        r++;
    }

    return r;
}

int bit_reverse(int x, int bits)
{
    int rev = 0;

    for (int j = 0; j < bits; j++) {
        rev = (rev << 1) | (x & 1);
        x = x >> 1;
    }

    return rev;
}

void bit_reverse_array(int input[], Complex output[], int n, int bits)
{
    for (int i = 0; i < n; i++) {
        int rev = bit_reverse(i, bits);
        output[rev] = (Complex){.real = input[i], .imag = 0};
    }
}

void fft(Complex x[], Complex twiddles[], int n, int bits) {
    for (int stage = 1; stage < bits + 1; stage++) {
        int m = 1 << stage;
        int half = m / 2;

        Complex w_m = twiddles[stage - 1];

        for (int base = 0; base < n; base += m) {
            Complex w = (Complex){.real = 1 << SCALE, .imag = 0};

            for (int k = 0; k < half; k++) {
                Complex t = complex_mult(w, x[base + k + half]);
                Complex u = x[base + k];

                x[base + k] = complex_add(u, t);
                x[base + k + half] = complex_sub(u, t);

                w = complex_mult(w, w_m);
            }
        }
    }
}
